crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,957 @@
+                self.subscriptions.lock_ref().to_owned(),
+            )))
+            .to_owned();
+
+        if let Some(ref h) = self.homeserver {
+            builder.homeserver(h.clone());
+        }
+        builder
+    }
+
+    /// Subscribe to a given room.
+    ///
+    /// Note: this does not cancel any pending request, so make sure to only
+    /// poll the stream after you've altered this. If you do that during, it
+    /// might take one round trip to take effect.
+    pub fn subscribe(&self, room_id: OwnedRoomId, settings: Option<v4::RoomSubscription>) {
+        self.subscriptions.lock_mut().insert_cloned(room_id, settings.unwrap_or_default());
+    }
+
+    /// Unsubscribe from a given room.
+    ///
+    /// Note: this does not cancel any pending request, so make sure to only
+    /// poll the stream after you've altered this. If you do that during, it
+    /// might take one round trip to take effect.
+    pub fn unsubscribe(&self, room_id: OwnedRoomId) {
+        if self.subscriptions.lock_mut().remove(&room_id).is_some() {
+            self.unsubscribe.lock_mut().push_cloned(room_id);
+        }
+    }
+
+    /// Lookup a specific room
+    pub fn get_room(&self, room_id: OwnedRoomId) -> Option<SlidingSyncRoom> {
+        self.rooms.lock_ref().get(&room_id).cloned()
+    }
+
+    /// Lookup a set of rooms
+    pub fn get_rooms<I: Iterator<Item = OwnedRoomId>>(
+        &self,
+        room_ids: I,
+    ) -> Vec<Option<SlidingSyncRoom>> {
+        let rooms = self.rooms.lock_ref();
+        room_ids.map(|room_id| rooms.get(&room_id).cloned()).collect()
+    }
+
+    async fn handle_response(
+        &self,
+        resp: v4::Response,
+        views: &[SlidingSyncView],
+    ) -> anyhow::Result<UpdateSummary> {
+        self.client.process_sliding_sync(resp.clone()).await?;
+        tracing::info!("main client processed.");
+        self.pos.replace(Some(resp.pos));
+        let mut updated_views = Vec::new();
+        if resp.lists.len() != views.len() {
+            bail!("Received response for {} lists, yet we have {}", resp.lists.len(), views.len());
+        }
+
+        for (view, updates) in views.iter().zip(resp.lists.iter()) {
+            let count: u32 =
+                updates.count.try_into().expect("the list total count convertible into u32");
+            tracing::trace!("view {:?}  update: {:?}", view.name, !updates.ops.is_empty());
+            if !updates.ops.is_empty() && view.handle_response(count, &updates.ops)? {
+                updated_views.push(view.name.clone());
+            }
+        }
+
+        let mut rooms = Vec::new();
+        let mut rooms_map = self.rooms.lock_mut();
+        for (id, room_data) in resp.rooms.iter() {
+            if let Some(mut r) = rooms_map.remove(id) {
+                r.update(room_data);
+                rooms_map.insert_cloned(id.clone(), r);
+                rooms.push(id.clone());
```

Comment:
I guess we remove and re-insert because [`MutableBTreeMapLockMut`](https://docs.rs/futures-signals/latest/futures_signals/signal_map/struct.MutableBTreeMapLockMut.html#) doesn't provide an `update` method? `MapDiff` has it so maybe it's just a PR away?